flake.nix

```diff
@@ -0,0 +1,28 @@
+{
+  inputs = {
+    nixpkgs.url = "github:NixOS/nixpkgs/nixos-23.05";
+    flake-utils.url = "github:numtide/flake-utils";
+  };
+
+  outputs = { self, nixpkgs, flake-utils }:
+    flake-utils.lib.eachDefaultSystem
+      (system: let
+        pkgs = import nixpkgs { system = "${system}"; config.allowUnfree = true; };
+      in
+      {
+        devShell = pkgs.mkShell { 
+          buildInputs = [ 
+            pkgs.figlet
+
+            pkgs.ruff
```

Comment:
Why is `ruff` different from the other dependencies?

Reply:
ruff is written in rust and does not depend on the environments version of python to run.

Since autopep8 and pytest require a python runtime we specify which runtime with the`withPackges` utility function. It basically ensures that the correct python version is used for the packages.

We could also specify `pkgs.python311 + pkgs.python311Packages.autopep8 ...` instead